docs/src/examples/autotuning-ridge.jl

```diff
@@ -0,0 +1,194 @@
+# # Auto-tuning Hyperparameters
+
+#md # [![](https://img.shields.io/badge/GitHub-100000?style=for-the-badge&logo=github&logoColor=white)](@__REPO_ROOT_URL__/examples/autotuning-ridge.jl)
+#md # [![](https://img.shields.io/badge/show-nbviewer-579ACA.svg)](@__NBVIEWER_ROOT_URL__/generated/autotuning-ridge.ipynb)
+
+# This example shows how to learn the hyperparameters in Ridge Regression using a gradient descent routine.  
+# Let the problem be modelled as
+
+# ```math
+# \begin{equation}
+# \min_{w} \quad \frac{1}{2n} \sum_{i=1}^{n} (y_{i} - w^T x_{i})^2 + \alpha \| w \|_2^2
+# \end{equation}
+# ```
+
+# where 
+# - `x`, `y` are the data points
+# - `w` constitutes weights of the regressing line
+# - `α` is the only hyperparameter acting on regularization
+
+using DiffOpt
+using Statistics
+using OSQP
+using JuMP
+using Plots
+import Random
+using LinearAlgebra
+
+
+"""
+Return the coefficient of determination R2 of the prediction.
+Best possible score is 1.0, it can be negative because the model can be arbitrarily worse
+"""
+function R2(y_true, y_pred)
+    u = sum((y_pred - y_true).^2)  # Regression sum of squares
+    v = sum((y_true .- mean(y_true)).^2)  # Total sum of squares
+    
+    return 1-(u/v)
+end
+
+# Create a non-trivial, noisy regression dataset
+
+function create_problem(N, D, noise)
+    w = rand(D) 
+    X = rand(N, D) 
+    
+    Y = X*w .+ noise*randn(N) # if noise=0, then there is no need of regularization and alpha=0 will give the best R2 score
+
+    l = N ÷ 2  # test train split
+    return X[1:l, :], X[l+1:N, :], Y[1:l], Y[l+1:N]
+end
+
+X_train, X_test, Y_train, Y_test = create_problem(800, 30, 4);
+
+
+# Define a helper function for regression
+
+function fit_ridge(X, Y, α)
+    model = Model(() -> diff_optimizer(OSQP.Optimizer))
+
+    N, D = size(X)
+    
+    @variable(model, w[1:D]>=-10)
+    set_optimizer_attribute(model, MOI.Silent(), true)
+    
+    @objective(
+        model,
+        Min,
+        sum((Y - X*w).*(Y - X*w))/(2.0*N) + α*(w'w),
```

Comment:
You take a scalar product in two different ways here. Doing `dot(Y - X*w, Y - X*w)` and `dot(w, w)` might be clearer